src/Http/WebUtilities/src/HttpRequestStreamReader.cs

```diff
@@ -258,45 +281,47 @@ public override async Task<int> ReadAsync(char[] buffer, int index, int count)
                         // _isBlocked == whether we read fewer bytes than we asked for.
                         _isBlocked = (_bytesRead < _byteBufferSize);
 
-                        Debug.Assert(n == 0);
+                        Debug.Assert(charsAvailable == 0);
 
                         _charBufferIndex = 0;
-                        n = _decoder.GetChars(
+                        charsAvailable = _decoder.GetChars(
                             _byteBuffer,
                             0,
                             _bytesRead,
                             _charBuffer,
                             0);
 
-                        Debug.Assert(n > 0);
+                        Debug.Assert(charsAvailable > 0);
 
-                        _charsRead += n; // Number of chars in StreamReader's buffer.
+                        _charsRead += charsAvailable; // Number of chars in StreamReader's buffer.
                     }
-                    while (n == 0);
+                    while (charsAvailable == 0);
 
-                    if (n == 0)
+                    if (charsAvailable == 0)
                     {
                         break; // We're at EOF
                     }
                 }
 
                 // Got more chars in charBuffer than the user requested
-                if (n > count)
+                if (charsAvailable > count)
                 {
-                    n = count;
+                    charsAvailable = count;
                 }
 
-                Buffer.BlockCopy(
-                    _charBuffer,
-                    _charBufferIndex * 2,
-                    buffer,
-                    (index + charsRead) * 2,
-                    n * 2);
+                var source = new Memory<char>(_charBuffer, _charBufferIndex, charsAvailable);
+                source.CopyTo(buffer);
 
-                _charBufferIndex += n;
+                if (charsAvailable < count)
```

Comment:
Why can't this be the same check done in Read with: 
```c#
                if (count > 0)
                {
                    buffer = buffer.Slice(charsRemaining, count);
                }
```

Reply:
aligned to `Read`. Also renamed `charsAvailable` to `charsRemaining` for simmetry.